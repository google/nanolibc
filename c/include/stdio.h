//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_

#include <stdarg.h>
#include <sys/types.h>
#include <sys/_cdefs.h>

/* Some assume stdio.h defines NULL, etc. And SUSv2 is actually ok with
 * that: "Inclusion of the <stdio.h> header may also make visible all
 * symbols from <stddef.h>." */
#include <stddef.h>

__BEGIN_DECLS

#define EOF (-1)

typedef struct _FILE_DESCR FILE;

extern FILE *stdout;
extern FILE *stderr;

int vsnprintf(char *str, size_t size, const char *format, va_list ap);
int snprintf(char *str, size_t size, const char *format, ...)
    __attribute__((format(printf, 3, 4)));

int vsprintf(char *str, const char *format, va_list ap);
int sprintf(char *str, const char *format, ...)
    __attribute__((format(printf, 2, 3)));

int vfprintf(FILE *stream, const char *format, va_list ap);
int fprintf(FILE *stream, const char *format, ...)
    __attribute__((format(printf, 2, 3)));

int vprintf(const char *format, va_list ap);
int printf(const char *format, ...) __attribute__((format(printf, 1, 2)));

int fputc(int c, FILE *stream);
int fputs(const char *s, FILE *stream);
int putc(int c, FILE *stream);
int putchar(int c);
int puts(const char *s);

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream);

int fflush(FILE *stream);

int perror(const char *s);

/*
 * Published but functions NOT implemented!
 */
extern FILE *stdin;

#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

int scanf(const char *format, ...);
int sscanf(const char *str, const char *format, ...);
FILE *fopen(const char *path, const char *mode);
int fclose(FILE *fp);
char *fgets(char *s, int size, FILE *stream);
size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);
int fseek(FILE *stream, long offset, int whence);
long ftell(FILE *stream);
void rewind(FILE *stream);

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_
