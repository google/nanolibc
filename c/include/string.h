//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_STRING_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_STRING_H_

#include <sys/_cdefs.h>
#include <sys/types.h>

__BEGIN_DECLS

int memcmp(const void *s1, const void *s2, size_t n);
void *memcpy(void *dstpp, const void *srcpp, size_t n);
void *memmove(void *dstpp, const void *srcpp, size_t n);
void *memset(void *dstpp, int c, size_t n);
void *memchr(const void *spp, int c, size_t n);
const void *rawmemchr(const void *spp, int c);

int strcmp(const char *p1, const char *p2);
char *strcpy(char *dest, const char *src);
size_t strlen(const char *s);
int strncmp(const char *s1, const char *s2, size_t n);
char *strncpy(char *dest, const char *src, size_t n);
char *strchr(const char *s, int i);
char *strrchr(const char *s, int i);
const char *strstr(const char *haystack, const char *needle);
size_t strspn(const char *s, const char *accept);
const char *strpbrk(const char *s, const char *accept);

// These two are unsupported.
char *strtok_r(char *str, const char *delim, char **saveptr);
char *strtok(char *str, const char *delim);

char *strdup(const char *s);
char *strndup(const char *s, size_t n);

char *strerror(int errnum);

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_STRING_H_
