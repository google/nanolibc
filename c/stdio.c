//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
 * Implementation of stdio.h functions.
 */

#include <stddef.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <stdio.h>

#include "third_party/nanolibc/c/libc_internals.h"
#include "third_party/tinyprintf/tinyprintf.h"

FILE *stdin = &__nnlc_internal_data.libc_stdin;
FILE *stdout = &__nnlc_internal_data.libc_stdout;
FILE *stderr = &__nnlc_internal_data.libc_stderr;

/* returns TRUE if param 'stream' is a valid FILE* pointer */
inline static int _is_valid_FILE(FILE *stream) {
  return ((stream != NULL) && (stream->magic == _NNLC_STDIO_MAGIC));
}

int fputc(int c, FILE *stream) {
  const unsigned char uc = c; /* according to man */
  if (!_is_valid_FILE(stream) || !stream->write) return EOF;
  if (1 != stream->write(&uc, 1)) return EOF;
  return uc; /* according to man */
}

int fputs(const char *s, FILE *stream) {
  size_t len;
  if (!_is_valid_FILE(stream) || !stream->write) return EOF;
  len = strlen(s);
  if (stream->write(s, len) != (ssize_t)len) return EOF;
  return 1;
}

int putc(int c, FILE *stream) { return fputc(c, stream); }

int putchar(int c) {
  const unsigned char uc = c; /* according to man */
  if (1 != stdout->write(&uc, 1)) return EOF;
  return uc; /* according to man */
}

int puts(const char *s) {
  const size_t len = strlen(s);
  if (stdout->write(s, len) != (ssize_t)len) return EOF;
  if (1 != stdout->write("\n", 1)) return EOF;
  return 1;
}

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream) {
  size_t i;
  if (!_is_valid_FILE(stream) || !stream->write) return 0;
  for (i = 0; i < nmemb; ++i) {
    const void *p = ptr + (i * size);
    if (stream->write(p, size) != (ssize_t)size) break;
  }
  return i;
}

int fflush(FILE *stream) {
  if (!_is_valid_FILE(stream)) return -1;

  return 0;
}

int perror(const char *s) {
  fprintf(stderr, "Error: %s\n", s);
  return 0;
}

/*
 * vprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to call the underlying runtime 'write'
 * functions. We use a small buffer on the stack to avoid calling the
 * underlying 'write' too often.
 */
struct _vprintf_putcf_data {
  FILE *stream;

  char *buff; /* allocated on the stack by vprintf */
  size_t buff_capacity;
  size_t nbuffered; /* slightly redundant with buff_capacity, we don't care */

  size_t total_written;
};

/* Callback function for tfp_format to add the given char c to the buffer */
static void _vprintf_putcf(void *p, char c) {
  struct _vprintf_putcf_data *data = (struct _vprintf_putcf_data *)p;

  /* sanity check */
  if (data->nbuffered >= data->buff_capacity)
    abort(); /* no assert(), to prevent infinite recursion */

  data->buff[data->nbuffered++] = c;

  /* need to flush buffer? */
  if (data->nbuffered >= data->buff_capacity) {
    int n = data->stream->write(data->buff, data->nbuffered);
    if (n > 0) data->total_written += n;
    data->nbuffered = 0;
  }
}

int vfprintf(FILE *stream, const char *format, va_list ap) {
  char buff[64]; /* any size >= 1 is Ok */
  struct _vprintf_putcf_data data;

  if (!_is_valid_FILE(stream) || !stream->write) return -1;

  data.stream = stream;
  data.buff_capacity = sizeof(buff);
  data.buff = buff;
  data.nbuffered = 0;
  data.total_written = 0;
  tfp_format(&data, _vprintf_putcf, format, ap);

  /* chars remaining in buffer? */
  if (data.nbuffered > 0) {
    int n = stream->write(data.buff, data.nbuffered);
    if (n > 0) data.total_written += n;
  }

  return data.total_written;
}

int fprintf(FILE *stream, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vfprintf(stream, format, ap);
  va_end(ap);
  return retval;
}

int vprintf(const char *format, va_list ap) {
  return vfprintf(stdout, format, ap);
}

int printf(const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vprintf(format, ap);
  va_end(ap);

  return retval;
}

/*
 * vsnprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to fill the buffer provided by caller.
 */
struct _vsnprintf_putcf_data {
  size_t dest_capacity;
  char *dest;
  size_t dest_used;
};

static void _vsnprintf_putcf(void *p, char c) {
  struct _vsnprintf_putcf_data *data = (struct _vsnprintf_putcf_data *)p;
  if (data->dest_used < data->dest_capacity) data->dest[data->dest_used] = c;

  data->dest_used++;
}

int vsnprintf(char *str, size_t size, const char *format, va_list ap) {
  struct _vsnprintf_putcf_data data;

  data.dest = str;
  data.dest_capacity = size;
  data.dest_used = 0;
  tfp_format(&data, _vsnprintf_putcf, format, ap);

  if (data.dest_used < size)
    data.dest[data.dest_used] = '\0';
  else if (size > 0)
    data.dest[size - 1] = '\0';

  return data.dest_used;
}

int snprintf(char *str, size_t size, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vsnprintf(str, size, format, ap);
  va_end(ap);
  return retval;
}

/*
 * vsprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to fill the buffer provided by caller.
 */
struct _vsprintf_putcf_data {
  char *dest;
  size_t dest_used;
};

static void _vsprintf_putcf(void *p, char c) {
  struct _vsprintf_putcf_data *data = (struct _vsprintf_putcf_data *)p;
  data->dest[data->dest_used++] = c;
}

int vsprintf(char *str, const char *format, va_list ap) {
  struct _vsprintf_putcf_data data;
  data.dest = str;
  data.dest_used = 0;
  tfp_format(&data, _vsprintf_putcf, format, ap);
  data.dest[data.dest_used] = '\0';
  return data.dest_used;
}

int sprintf(char *str, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vsprintf(str, format, ap);
  va_end(ap);
  return retval;
}
