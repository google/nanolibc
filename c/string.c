//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
 * Implementation of libC string.h functions.
 */

#include <stddef.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

/* These are weak symbols, so that we continue using gnu-efi's version
 * when we link against it */
int memcmp(const void *p1, const void *p2, size_t n)
    __attribute__ ((weak));
void *memcpy(void *dstpp, const void *srcpp, size_t n)
    __attribute__ ((weak));
void *memmove(void *dstpp, const void *srcpp, size_t n)
    __attribute__ ((weak));
void *memset(void *dstpp, int c, size_t n)
    __attribute__ ((weak));

/* Highly NON optimized, but simple */
int memcmp(const void *p1, const void *p2, size_t n) {
  const char *s1 = p1;
  const char *s2 = p2;

  while (n > 0) {
    if (*s1 != *s2) return *s1 - *s2;

    s1++;
    s2++;
    n--;
  }

  return 0;
}

/* Highly NON optimized, but simple */
void *memcpy(void *dstpp, const void *srcpp, size_t n) {
  char *d = dstpp;
  const char *s = srcpp;

  while (n > 0) {
    *d++ = *s++;
    n--;
  }

  return dstpp;
}

/* Highly NON optimized, but simple */
void *memmove(void *dstpp, const void *srcpp, size_t n) {
  if (dstpp <= srcpp) {
    return memcpy(dstpp, srcpp, n);
  }
  char *d = dstpp + n;
  const char *s = srcpp + n;

  while (n > 0) {
    *--d = *--s;
    n--;
  }

  return dstpp;
}

/* Highly NON optimized, but simple */
void *memset(void *dstpp, int c, size_t n) {
  char *d = dstpp;

  while (n > 0) {
    *d++ = c;
    n--;
  }

  return dstpp;
}

const void *rawmemchr(const void *spp, int c) {
  const unsigned char *s = spp;

  while (*s != (unsigned char)c) s++;

  return s;
}

void *memchr(const void *spp, int c, size_t n) {
  const unsigned char *s = spp;

  while (n-- != 0) {
    if (*s == (unsigned char)c) {
      return (void*)s;
    }
    s++;
  }

  return NULL;
}

int strcmp(const char *p1, const char *p2) {
  const unsigned char *s1 = (const unsigned char *)p1;
  const unsigned char *s2 = (const unsigned char *)p2;

  while (1) {
    const unsigned char c1 = *s1++;
    const unsigned char c2 = *s2++;
    if ((c1 == '\0') || (c1 != c2)) return (c1 - c2);
  }

  /* not reached */
  return 0;
}

/* Highly NON optimized, but simple */
int strncmp(const char *p1, const char *p2, size_t n) {
  const unsigned char *s1 = (const unsigned char *)p1;
  const unsigned char *s2 = (const unsigned char *)p2;

  while (n > 0) {
    const unsigned char c1 = *s1++;
    const unsigned char c2 = *s2++;
    if ((c1 == '\0') || (c1 != c2)) return (c1 - c2);
    n--;
  }

  return 0;
}

char *strcpy(char *dest, const char *src) {
  char *d = dest;
  const char *s = src;

  while (*s) *d++ = *s++;

  *d = '\0';
  return dest;
}

/* strncpy borrowed from Linux man pages 3.35 */
char *strncpy(char *dest, const char *src, size_t n) {
  size_t i;

  for (i = 0; i < n && src[i] != '\0'; i++) dest[i] = src[i];
  for (; i < n; i++) dest[i] = '\0';

  return dest;
}

/* Highly NON optimized, but simple */
size_t strlen(const char *s) {
  size_t i;
  for (i = 0; *s; ++s, ++i) continue;
  return i;
}

char *strchr(const char *s, int i) {
  for (; *s; ++s)
    if (*s == i) return (char *)s;

  return NULL;
}

char *strrchr(const char *s, int i) {
  const char *last = NULL;

  for (; *s; ++s)
    if (*s == i) last = s;

  return (char *)last;
}

const char *strstr(const char *haystack, const char *needle) {
  const char *h = haystack;
  const char *n = needle;
  const char *found = NULL;

  if (*needle == '\0') return haystack;

  while (*h) {
    if ((found != NULL) && (*h != *n)) {
      found = NULL;
      n = needle;
    }

    if (*h == *n) {
      if (found == NULL) found = h;
      n++;
      if (*n == '\0') return found;
    }

    h++;
  }

  return NULL;
}

/* strspn borrowed from eglibc 2.17 */
size_t strspn(const char *s, const char *accept) {
  const char *p;
  const char *a;
  size_t count = 0;

  for (p = s; *p != '\0'; ++p) {
    for (a = accept; *a != '\0'; ++a)
      if (*p == *a) break;
    if (*a == '\0')
      return count;
    else
      ++count;
  }

  return count;
}

/* Naive O(M*N) implementation, good enough if "accept" is small */
const char *strpbrk(const char *s, const char *accept) {
  while (*s != '\0') {
    const char *a = accept;
    while (*a != '\0')
      if (*a++ == *s) return s;
    ++s;
  }

  return NULL;
}

/* strtok_r borrowed from eglibc 2.17's strtok */
char *strtok_r(char *s, const char *delim, char **saveptr) {
  char *token;

  if (s == NULL) s = *saveptr;

  /* Scan leading delimiters.  */
  s += strspn(s, delim);
  if (*s == '\0') {
    *saveptr = s;
    return NULL;
  }

  /* Find the end of the token.  */
  token = s;
  s = (char *)strpbrk(token, delim);
  if (s == NULL) {
    /* This token finishes the string.  */
    *saveptr = (char *)rawmemchr(token, '\0');
  } else {
    /* Terminate the token and make *saveptr point past it.  */
    *s = '\0';
    *saveptr = s + 1;
  }
  return token;
}

char *strtok(char *s, const char *delim) {
  static char *global_saveptr = NULL;
  return strtok_r(s, delim, &global_saveptr);
}

char *strdup(const char *s) {
  const size_t len = strlen(s);
  char *result = malloc(len+1);
  if (NULL != result) {
    strcpy(result, s);
  }
  return result;
}

char *strndup(const char *s, size_t n) {
  size_t len = strlen(s);
  if (n < len) {
    len = n;
  }
  char *result = malloc(len+1);
  if (NULL != result) {
    memcpy(result, s, len);
    result[len] = '\0';
  }
  return result;
}

int errno = 0; /* NOT MT-aware/MT-safe */

char *strerror(int errnum) {
  if (errnum == 0) return "no error";
  return "unexpected error";
}
