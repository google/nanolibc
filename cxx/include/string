#ifndef THIRD_PARTY_NANOLIBC_CXX_INCLUDE_STRING_
#define THIRD_PARTY_NANOLIBC_CXX_INCLUDE_STRING_

#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

namespace std {

class string {
 public:
  static constexpr size_t npos = ~static_cast<size_t>(0);

  // Creates a string of `size` elements initilized to `value`.
  explicit string(size_t size = 0, char value = '\0')
      : size_(size), data_(NewData(size_)) {
    if (size_ != 0) {
      memset(data_, value, size_);
    }
  }

  // Creates a string of `size` bytes from `data`.
  string(const void* data, size_t size) : size_(size), data_(NewData(size_)) {
    if (size_ != 0) {
      memcpy(data_, data, size_);
    }
  }

  // Create a string from a '\0' terminated C string.
  string(const char* data) : size_(strlen(data)), data_(NewData(size_)) {
    if (size_ != 0) {
      memcpy(data_, data, size_);
    }
  }

  // Copy constructor.
  string(const string& other) : size_(other.size_), data_(NewData(size_)) {
    if (size_ != 0) {
      memcpy(data_, other.data_, size_);
    }
  }

  // Copy constructor.
  string(string&& other) : size_(other.size_), data_(other.data_) {
    other.data_ = nullptr;
    other.size_ = 0;
  }

  // Move operator.
  string& operator= (string&& other) {
    clear();
    size_ = other.size_;
    data_ = other.data_;
    other.data_ = nullptr;
    other.size_ = 0;
    return *this;
  }

  // Assignment operator.
  string& operator= (const string& other) {
    clear();
    size_ = other.size_;
    data_ = NewData(size_);
    if (size_ != 0) {
      memcpy(data_, other.data_, size_);
    }
    return *this;
  }

  ~string() { clear(); }

  string substr(size_t pos = 0, size_t len = ((size_t)-1)) const {
    // Note: be careful about checking for overflow.
    if (pos > size_) {
      fprintf(stderr, "Indexed past end of string buffer in substr.\n");
      abort();
    }
    if (len > size_ - pos) {
      len = size_ - pos;
    }
    return string(data_ + pos, len);
  }

  size_t size() const { return size_; }
  bool empty() const { return size_ == 0; }
  void clear() {
    delete[] data_;
    data_ = nullptr;
    size_ = 0;
  }

  char& operator[](size_t index) { return data_[index]; }
  char operator[](size_t index) const { return data_[index]; }
  char& at(size_t index) { return data_[index]; }
  char at(size_t index) const { return data_[index]; }

  char* data() const {
    if (data_ == nullptr) {
      return "";
    }
    return data_;
  }
  const char* c_str() const { return data(); }

  int compare(char* data, size_t size) const {
    size_t min = size_ <= size ? size_ : size;
    int res = 0;
    if (min != 0) {
      res = memcmp(data_, data, min);
    }
    if (res != 0) {
      return res;
    }
    if (size_ < size) {
      return -1;
    }
    if (size_ > size) {
      return 1;
    }
    return 0;
  }

  int compare(const string& other) const {
    return compare(other.data_, other.size_);
  }

  bool operator==(const string& other) const { return compare(other) == 0; }
  bool operator!=(const string& other) const { return compare(other) != 0; }
  bool operator<=(const string& other) const { return compare(other) <= 0; }
  bool operator>=(const string& other) const { return compare(other) >= 0; }
  bool operator<(const string& other) const { return compare(other) < 0; }
  bool operator>(const string& other) const { return compare(other) > 0; }

  void resize(size_t size, char c = '\0') {
    char* data = NewData(size);
    size_t min_pos = size <= size_ ? size : size_;
    if (min_pos != 0) {
      memcpy(data, data_, min_pos);
    }
    size_t num_to_pad = size - min_pos;
    if (num_to_pad != 0) {
      memset(data + min_pos, c, num_to_pad);
    }
    clear();
    data_ = data;
    size_ = size;
  }

  string& operator+=(const string& other) {
    if (other.size_ != 0) {
      size_t old_size = size_;
      resize(size_ + other.size_);
      memcpy(data_ + old_size, other.data_, other.size_);
    }
    return *this;
  }

  string& append(const string& other) {
    size_t old_size = size_;
    resize(size_ + other.size_);
    memcpy(data_ + old_size, other.data_, other.size_);
    return *this;
  }

  string& replace(size_t pos, size_t len, const string& str) {
    if (pos > size_) {
      fprintf(stderr, "Indexed past end of string buffer in replace.\n");
      abort();
    }
    if (len > size_ - pos) {
      len = size_ - pos;
    }
    size_t new_size = size_ + str.size_ - len;
    if (new_size == 0) {
      clear();
      return *this;
    }
    char* new_data = NewData(new_size);
    memcpy(new_data, data_, pos);
    memcpy(new_data + pos, str.data_, str.size_);
    size_t remaining = size_ - pos - len;
    memcpy(new_data + pos + str.size_, data_ + pos + len, remaining);
    delete[] data_;
    data_ = new_data;
    size_ = new_size;
    return *this;
  }

  char* begin() const { return data_; }
  char* end() const { return data_ + size_; }

 private:
  // Like all other std::string implementations, we allocate a '\0' at the end
  // of every non-empty string so that c_str() can just return data().
  static char* NewData(size_t size) {
    if (size == 0) {
      return nullptr;
    }
    char* data = new char[size + 1];
    data[size] = '\0';
    return data;
  }

  size_t size_;
  char* data_;
};

inline bool operator==(const char* data, const string& s) {
  return s.compare(data) == 0;
}
inline bool operator!=(const char* data, const string& s) {
  return s.compare(data) != 0;
}
inline bool operator<=(const char* data, const string& s) {
  return s.compare(data) >= 0;
}
inline bool operator>=(const char* data, const string& s) {
  return s.compare(data) <= 0;
}
inline bool operator<(const char* data, const string& s) {
  return s.compare(data) > 0;
}
inline bool operator>(const char* data, const string& s) {
  return s.compare(data) < 0;
}

inline string operator+(const string& s1, const string& s2) {
  string result(s1.size() + s2.size());
  memcpy(result.data(), s1.data(), s1.size());
  memcpy(result.data() + s1.size(), s2.data(), s2.size());
  return result;
}

inline string to_string(size_t val) {
  char buf[32];  // Large enough to hold 1 << 64 in base 10.
  sprintf(buf, "%u", val);
  return buf;
}

}  // namespace std

#endif  // THIRD_PARTY_NANOLIBC_CXX_INCLUDE_STRING_
