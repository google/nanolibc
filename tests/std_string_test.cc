//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* Unit tests for std::string */

#include <stdio.h>
#include <string.h>

#include <string>
#include <utility>

#include "third_party/nanolibc/tests/test_utils/nnlc_test.h"

namespace {

// Just test we can assign Hello, World!.
void TestAssign() {
  std::string s1 = "Hello, World!";
  std::string s2 = s1;
  ASSERT(s1 == s2);
  ASSERT(s1 == "Hello, World!");
  std::string s3(4, 'x');
  printf("TestAssign passed\n");
}

// Test constructors.
void TestConstructors() {
  // Test default constructor.
  std::string s;
  ASSERT(s.empty());
  ASSERT(s == "");        // NOLINT
  ASSERT(s.size() == 0);  // NOLINT
  const char kData[] = "xxxx";
  std::string s1(4, 'x');
  ASSERT(s1.size() == 4);
  ASSERT(s1 == kData);
  ASSERT(!strcmp(s1.c_str(), kData));
  std::string s2(kData, sizeof(kData) - 1);
  ASSERT(s2 == kData);
  const char kLongData[] = "xxxx\x00";
  // Construction from C strings look for a terminating '\0'.
  std::string temp(kLongData);
  ASSERT(s2 == kLongData);
  const char kHello[] = "Constructor test passed";
  std::string s3(kHello);
  ASSERT(s3 == kHello);
  std::string s4(s3);
  ASSERT(s4 == s3);
  std::string s5 = s4;
  ASSERT(s5 == kHello);
  // Test copy constructor.
  std::string s6(s5);
  ASSERT(s6 == kHello);
  ASSERT(s6.data() != s5.data());
  // NOLINTBEGIN
  // Test move constructor.
  std::string s7 = std::move(s6);
  ASSERT(s7 == kHello);
  ASSERT(s6.empty());
  // NOLINTEND
  // Record in golden test that printing c_str() does not change.
  printf("%s\n", s3.c_str());
  printf("TesstConstructors passed\n");
}

// Test various methods of std::string.
void TestMethods() {
  std::string s1("test");
  ASSERT(!s1.empty());
  ASSERT(s1.substr(2, 2) == "st");
  ASSERT(s1.substr(4, 100).empty());
  s1.clear();
  ASSERT(s1.empty());
  std::string s2("Hello, World!");
  ASSERT(!strcmp(s2.data(), s2.c_str()));
  ASSERT(strlen(s2.data()) == s2.size());
  std::string s3("Hello, ");
  s3.append("World!");
  ASSERT(s3 == "Hello, World!");
  ASSERT(s3.size() ==
         std::string("Hello, ").size() + std::string("World!").size());
  printf("TestMethods passed\n");
}

// Test string indexing operations.
void TestIndexing() {
  std::string s1("Hello, World!");
  ASSERT(s1[0] == 'H' && s1[s1.size() - 1] == '!');
  // Violating the abstraction here: we know there is a trailing '\0' in this
  // implementation.
  ASSERT(s1[s1.size()] == '\0');
  s1[0] = 'h';
  s1[7] = 'w';
  ASSERT(s1 == "hello, world!");
  ASSERT(s1.at(0) == 'h');
  ASSERT(s1.at(1) == 'e');
  ASSERT(s1.at(7) == 'w');
  s1.at(s1.size() - 1) = '.';
  ASSERT(s1 == "hello, world.");
  printf("TestIndexing passed\n");
}

// Test relational operators
void TestRelationalOperators() {
  // From smallest to largest.
  std::string s1("a");
  std::string s2("aa");
  std::string s3("b");
  std::string s4("b\x00", 2);
  ASSERT(s1 == s1);
  ASSERT(s1 != s2);
  ASSERT(s1 < s2);
  ASSERT(s2 < s3);
  ASSERT(s3 < s4);
  ASSERT(s4 > s3);
  ASSERT(s3 > s2);
  ASSERT(s2 > s1);
  ASSERT(s1 <= s2);
  ASSERT(s2 <= s3);
  ASSERT(s3 <= s4);
  ASSERT(s4 >= s3);
  ASSERT(s3 >= s2);
  ASSERT(s2 >= s1);
  ASSERT(s1 <= s1);
  ASSERT(s1 >= s1);
  const char cstr1[] = "aa";
  ASSERT(cstr1 == s2);
  ASSERT(cstr1 != s3);
  ASSERT(cstr1 <= s2);
  ASSERT(cstr1 <= s3);
  ASSERT(cstr1 >= s2);
  ASSERT(cstr1 >= s1);
  ASSERT(cstr1 < s3);
  ASSERT(cstr1 > s1);
  printf("TestRelationalOperators passed\n");
}

// Test the resize method.
void TestResize() {
  std::string s;
  s.resize(4);
  ASSERT(s == std::string(4, '\0'));
  s[0] = 'x';
  s[1] = 'x';
  s[2] = 'x';
  s[3] = 'x';
  s.resize(6);
  ASSERT(s == std::string("xxxx\x00\x00", 6));
  s.resize(3);
  ASSERT(s == std::string("xxx", 3));
  s.resize(4);
  ASSERT(s == std::string("xxx\x00", 4));
  printf("TestResize passed\n");
}

// Test the addition operator.
void TestAdditionOperator() {
  ASSERT(std::string("a") + "b" == "ab");
  // From smallest to largest.
  std::string a("a");
  std::string b("aa");
  std::string c("b");
  std::string d("b\x00", 2);
  ASSERT(a + b == "aaa");
  ASSERT(a + b + c == "aaab");
  ASSERT(a + b + c + d == std::string("aaabb\x00", 6));
  std::string sum;
  ASSERT(sum.empty());
  sum += a;
  sum += b;
  sum += c;
  sum += d;
  ASSERT(sum == a + b + c + d);
  std::string nothing;
  ASSERT(nothing.empty());
  ASSERT(nothing + nothing == nothing);
  ASSERT(a + nothing == a);
  ASSERT(nothing + a == a);
  ASSERT("Hello, " + std::string("World!") == "Hello, World!");
  printf("TestAdditionOperator passed\n");
}

// Test iterating over string characters.
void TestIterators() {
  std::string s("Hello, World!");
  size_t i = 0;
  for (char c : s) {
    ASSERT(c == s[i]);
    i++;
  }
  ASSERT(i == s.size());
  auto it = s.begin();
  i = 0;
  while (it != s.end()) {
    ASSERT(*it == s[i]);
    it++;
    i++;
  }
  ASSERT(static_cast<size_t>(it - s.begin()) == s.size());
  printf("TestIterators passed\n");
}

// Since C++11, the string::data() returned pointer can be written to legally.
void TestWriteData() {
  std::string s = "test";
  char* p = s.data();
  p[0] = 'T';
  p[3] = 'T';
  ASSERT(s == "TesT");
  printf("TesstWriteData  passed\n");
}

// Test the replace method.  It must behave well even when an attacker controls
// the inputs.
void TestReplace() {
  std::string buf("hello");
  std::string& result = buf.replace(0, 1, "H");
  ASSERT(buf == "Hello");
  // Make sure we returned *this.
  ASSERT(result.data() == buf.data());
  buf.replace(5, std::string::npos, ", World!");
  ASSERT(buf == "Hello, World!");
  buf.replace(7, 1, std::string("great big w"));
  ASSERT(buf == "Hello, great big world!");
  buf.replace(7, 10, "");
  ASSERT(buf == "Hello, world!");
  buf.replace(6, 0, "");
  ASSERT(buf == "Hello, world!");
  buf.replace(0, 0, "!");
  ASSERT(buf == "!Hello, world!");
  printf("TestReplace passed\n");
}

// Test any supported std::to_string() functions.
void TestToString() {
  ASSERT(std::to_string(0xdeadbeef) == "3735928559");
  printf("TestToString passed\n");
}

}  // namespace

int main() {
  TestAssign();
  TestConstructors();
  TestMethods();
  TestResize();
  TestRelationalOperators();
  TestAdditionOperator();
  TestIndexing();
  TestIterators();
  TestWriteData();
  TestReplace();
  TestToString();
  printf("PASSED\n");
  return 0;
}
