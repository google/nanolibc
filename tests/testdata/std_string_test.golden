TestAssign passed
Constructor test passed
TesstConstructors passed
TestMethods passed
TestResize passed
TestRelationalOperators passed
TestAdditionOperator passed
TestIndexing passed
TestIterators passed
TesstWriteData  passed
TestReplace passed
TestToString passed
PASSED
# END TEST WITH RETVAL=0
